docs/usage.rst

```diff
@@ -1182,6 +1182,31 @@ function. In other words, apply or check units at interfaces, but during an
 internal calculation it is often worth stripping units, especially if the
 calculation involves many operations on arrays with only a few elements.
 
+:class:`unyt_array.name <unyt.array.unyt_array.name>` attribute
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+
+The unyt_array has a name attribute for use in structured-data applications or
+similar applications that require labeled data. For example, Numpy has record arrays
+and when constructed as shown below, it is possible to retain the units while taking
+advantage of the labeled record fields.
+
+  >>> import numpy as np
+  >>> from unyt import unyt_array
+  >>> x = unyt_array([0, 1, 2], "s", name="time")
+  >>> y = unyt_array([3, 4, 5], "m", name="distance")
+  >>> data = (x, y)
+  >>> dt = [(a.name, "O") for a in data]
+  >>> data_points = np.array(list(zip(*data)), dtype=dt).view(np.recarray)
+  >>> data_points[0].time
+  unyt_quantity(0, 's')
+  >>> data_points[0].distance
+  unyt_quantity(3, 'm')
+
+.. note::
+  The name attribute does not propagate through mathematical operations or even the
+  unyt_array.copy() function. The only exception to this is the unyt_array.convert_to_units()
```

Comment:
Now that you mention it, I would actually make `copy` propagate that attribute, otherwise it might surprisingly not be a copy in just this one way.

Reply:
And maybe `in_units` and `to` as well?

Reply:
``to`` and ``in_units`` create a unyt_array that is missing the name attribute.
```python
>>> from unyt import s
>>> x = [0,1,2]*s
>>> hasattr(x, "name")
True
>>> y = x.to("ms")
>>> hasattr(y, "name")
False
```
Should I investigate?

Reply:
I updated ``copy`` to propagate ``name``. now,
```python
>>> from unyt import s
>>> x = [0,1,2]*s
>>> x.name = "time"
>>> y.copy()
>>> y.name
'time'
```

Reply:
I bet it has something to do with the way the returned array is created: https://github.com/yt-project/unyt/blob/595b9cdad094a9cd7f04980aee13a96344c0fc5f/unyt/array.py#L819

If I had to hazard a guess, `bypass_validation=True` makes it so you return from `__new__` before `name` gets set. Maybe `name` needs to get set earlier? `bypass_validation` is there for performance, in code where you’re already sure that what you’re passing in is an `ndarray` and a `Unit` object, you can set that to `True` and bypass a bunch of validation code which can substantially improve performance in hot code paths. It’s mostly an internal hack that enables some of the speed of `unyt`.

Reply:
correct. now ``to`` and ``in_units`` have the attribute. Should I propagate name for these two functions?

Reply:
Yes I think so. Just changing the units doesn’t really change the semantic meaning of the data. Doing a mathematical operation might, so we don’t propagate `name` in `__array_ufunc__`.